### ISSUE TITLE
Protocol method implemented by an instance callable is considered abstract

### ISSUE THREAD

**Issue by jelle-openai** [astral-sh/ty#4570]
### Summary

A concrete subclass assigns a callable to an instance attribute with the same name as a protocol method. ty 0.0.81 accepts this; ty 0.0.82 reports `call-non-callable` when constructing the subclass.

```python
from collections.abc import Callable
from typing import Protocol

class Capybara(Protocol):
    def forage(self) -> None: ...

class Hutia(Capybara):
    forage: Callable[[], None]

    def __init__(self) -> None:
        self.forage = lambda: None

Hutia().forage()
```

Expected: no diagnostic. The assignment implements the protocol method, and `Hutia().forage()` executes successfully in Python.

Actual with 0.0.82:

```text
error[call-non-callable]: Cannot instantiate `Hutia` with unimplemented abstract method `forage`
```

0.0.81: `All checks passed!`

The same behavior occurs with an async protocol method and an `AsyncMock`-typed instance attribute assigned in `__init__`.

Possibly related to [#4552](https://github.com/astral-sh/ty/issues/4552), but this example has no slots or abstract property. That issue is closed following a slots-specific fix.

**Comment by AlexWaygood** [comment]
This class can be instantiated:

```pycon
>>> from collections.abc import Callable
... from typing import Protocol
...
... class Capybara(Protocol):
...     def forage(self) -> None: ...
...
... class Hutia(Capybara):
...     forage: Callable[[], None]
...
...     def __init__(self) -> None:
...         self.forage = lambda: None
...
... Hutia().forage()
...
>>>
```

But if you explicitly decorate the protocol method with `@abstractmethod`, it cannot:

```pycon
>>> from collections.abc import Callable
... from typing import Protocol
... from abc import abstractmethod
...
... class Capybara(Protocol):
...     @abstractmethod
...     def forage(self) -> None: ...
...
... class Hutia(Capybara):
...     forage: Callable[[], None]
...
...     def __init__(self) -> None:
...         self.forage = lambda: None
...
... Hutia().forage()
...
Traceback (most recent call last):
  File "<python-input-1>", line 15, in <module>
    Hutia().forage()
    ~~~~~^^
TypeError: Can't instantiate abstract class Hutia without an implementation for abstract method 'forage'
```

So if we wanted to fix this, we would have to attempt to differentiate between explicitly abstract methods and implicitly abstract methods on protocols. I'm not totally sure if that would be worth it? It feels like it could cause confusing inconsistencies elsewhere.

**Comment by AlexWaygood** [comment]
There's also the issue that certain protocol methods _cannot_ be soundly implemented by instance attributes (any dunder method that must be looked up on the class rather than the instance, for example)

**Comment by carljm** [comment]
All other type checkers avoid this false positive by generally allowing instance attributes to implement abstract methods; of course this means they instead have false negatives in some cases that will raise at runtime.
